### PR TITLE
handle retries if possible when exceptions occur in calls to Azure

### PR DESCRIPTION
**Pull request by scotmoor**
Wrap both "get" and "list" Azure API calls in logic to handle both socket timeouts and HTTP responses which can be retried.
Fixed bug in Server Group caching
Fixed bug in delete resource operation when checking for HTTP response code of 404

### CODE REVIEW

**Comment by scotmoor** [comment]
@JargoonPard @milismsft PTAL

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureComputeClient.groovy [hunk 1]
```diff
@@ -68,25 +67,27 @@ public class AzureComputeClient extends AzureBaseClient {
 
     try {
       // Usage of local variables to ease with debugging the code; keeping the content retrieved from Azure JSDK call to help with stepping through the code and inspect the values
-      List<String> publishers = client.getVirtualMachineImagesOperations()?.listPublishers(location)?.body?.collect { it.name }
+      def ops = client.getVirtualMachineImagesOperations()
```
Comment (milismsft) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureComputeClient.groovy]: About a month ago when I tried to reuse the ops object I was getting random timeouts/ops object not longer being valid. Is this fixed now? Can you please confirm that you get the expected resources in the redis cache (start with an empty cache).

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureComputeClient.groovy [hunk 2]
```diff
@@ -161,8 +164,7 @@ public class AzureComputeClient extends AzureBaseClient {
   }
 
   ServiceResponse<Void> destroyServerGroup(String resourceGroupName, String serverGroupName) {
-    VirtualMachineScaleSetsOperations ops = getAzureOps(
-      client.&getVirtualMachineScaleSetsOperations, "Get operations object", "Failed to get operation object") as VirtualMachineScaleSetsOperations
+    VirtualMachineScaleSetsOperations ops = client.getVirtualMachineScaleSetsOperations()
```
Comment (milismsft) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureComputeClient.groovy]: Did the Azure JSDK team confirm that getting the operation object will not result into a REST call to Azure?

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy [hunk 3]
```diff
@@ -63,9 +60,9 @@ class AzureNetworkClient extends AzureBaseClient {
     def result = new ArrayList<AzureLoadBalancerDescription>()
 
     try {
-      def loadBalancers = this.client.getLoadBalancersOperations().listAll().body
+      def loadBalancers = executeOp({client.getLoadBalancersOperations().listAll()})?.body
```
Comment (milismsft) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy]: Will there be any case where we could end-up with a null operation object?

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureBaseClient.groovy [hunk 4]
```diff
@@ -54,28 +54,52 @@ public abstract class AzureBaseClient {
...
+          return null
+        }
+        else {
           throw e
         }
       }
     }
-
-    result
+    null
```
Comment (ghost) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureBaseClient.groovy]: This shouldn't be reachable.

Comment (scotmoor) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureBaseClient.groovy]: Agreed but without it, we get a " Not all execution paths return a value" warning. If we are ok with the warning I will remove it.

Comment (ghost) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureBaseClient.groovy]: Not a big deal either way. Might as well just leave it.

**Comment by ghost** [comment]
LGTM

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy (hunk 3 above)
Comment (scotmoor) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy]: Only if an exception is thrown by the constructor

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureComputeClient.groovy (hunk 2 above)
Comment (scotmoor) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureComputeClient.groovy]: Yes. This method instantiates the class that declares the "operations" (APIs) that are available. It also does initialization of the retrofit client (which still doesn't make a rest call). Unfortunately, per the SDK team, the initialization of the retrofit client can be expensive and they have recommended that we limit our calls to instantiate the "operations" classes when possible.

**Comment by scotmoor** [comment]
@JargoonPard @milismsft 
Some updates based upon feedback from the Azure SDK team. PTAL

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureComputeClient.groovy [hunk 5]
```diff
@@ -45,6 +48,16 @@ public class AzureComputeClient extends AzureBaseClient {
...
+
+  @Lazy
+  private VirtualMachineImagesOperations vmImageOps = {client.getVirtualMachineImagesOperations()}()
+
+  @Lazy
+  private VirtualMachineScaleSetsOperations scaleSetOps = {client.getVirtualMachineScaleSetsOperations()}()
+
+  @Lazy
+  private VirtualMachineScaleSetVMsOperations scaleSetVMOps = {client.getVirtualMachineScaleSetVMsOperations()}()
+
```
Comment (milismsft) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureComputeClient.groovy:57]: What's the lifetime expectation for these objects? Is it reasonable to expect that they will never "expire"?

**Pull request closed without merging**